test/com/esotericsoftware/kryo/serializers/SerializerPriorityTest.java

```diff
@@ -0,0 +1,126 @@
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. */
+
+package com.esotericsoftware.kryo.serializers;
+
+import com.esotericsoftware.kryo.DefaultSerializer;
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.KryoSerializable;
+import com.esotericsoftware.kryo.Serializer;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import com.esotericsoftware.kryo.serializers.DefaultSerializers.KryoSerializableSerializer;
+import com.esotericsoftware.kryo.serializers.DefaultSerializers.TreeMapSerializer;
+import org.junit.Test;
+
+import java.util.TreeMap;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * @author Niels Basjes
+ */
+@SuppressWarnings("synthetic-access")
+public class SerializerPriorityTest {
+
+    // Dummy serializer
+    public static final class MyAnnotationSerializer extends Serializer<Object> {
+        @Override
+        public void write(Kryo kryo, Output output, Object object) {
+        }
+
+        @Override
+        public Object read(Kryo kryo, Input input, Class type) {
+            return null;
+        }
+    }
+
+    // Dummy KryoSerializable implementation
+    public interface BaseKryoSerializable extends KryoSerializable {
+        @Override
+        default void write(Kryo kryo, Output output) {
+        }
+
+        @Override
+        default void read(Kryo kryo, Input input) {
+        }
+    }
+
+    public static final class MyTreeMap extends TreeMap<String, String> {   }
+    public static final class MyKryoSerializable implements BaseKryoSerializable {    }
+    public static final class MyTreeMapKryoSerializable extends TreeMap<String, String> implements BaseKryoSerializable {    }
+
+    @DefaultSerializer(MyAnnotationSerializer.class)
+    public static final class MyAnnotated {    }
+
+    @DefaultSerializer(MyAnnotationSerializer.class)
+    public static final class MyAnnotatedTreeMap extends TreeMap<String, String> {    }
+
+    @DefaultSerializer(MyAnnotationSerializer.class)
+    public static final class MyAnnotatedKryoSerializable implements BaseKryoSerializable {    }
+
+    @DefaultSerializer(MyAnnotationSerializer.class)
+    public static final class MyAnnotatedTreeMapKryoSerializable extends TreeMap<String, String> implements BaseKryoSerializable {    }
+
+    private void checkCorrectSerializer(Class<?> myClass, Class<? extends Serializer<?>> expectedSerializer) {
```

Comment:
Please move this method to the end of the test class. I'd prefer the name `assertSerializer`

Reply:
Done